src/test/java/htsjdk/tribble/gff/Gff3WriterTest.java

```diff
@@ -0,0 +1,155 @@
+package htsjdk.tribble.gff;
+
+import htsjdk.HtsjdkTest;
+import htsjdk.tribble.AbstractFeatureReader;
+import htsjdk.tribble.TestUtils;
+import htsjdk.tribble.Tribble;
+import htsjdk.tribble.TribbleException;
+import htsjdk.tribble.readers.LineIterator;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+public class Gff3WriterTest extends HtsjdkTest {
+    final static String DATA_DIR = TestUtils.DATA_DIR + "/gff/";
+    private final Path ensembl_human_small = Paths.get(DATA_DIR + "Homo_sapiens.GRCh38.97.chromosome.1.small.gff3");
+    private final Path gencode_mouse_small = Paths.get(DATA_DIR + "gencode.vM22.annotation.small.gff3");
+    private final Path ncbi_woodpecker_small = Paths.get(DATA_DIR + "ref_ASM69900v1_top_level.small.gff3");
+    private final Path feature_extends_past_circular_region = Paths.get(DATA_DIR + "feature_extends_past_circular_region.gff3");
+    private final Path with_fasta = Paths.get(DATA_DIR + "fasta_test.gff3");
+    private final Path with_fasta_artemis = Paths.get(DATA_DIR + "fasta_test_artemis.gff3");
+    private final Path ordered_cofeature = Paths.get(DATA_DIR, "ordered_cofeatures.gff3");
+    private final Path child_before_parent = Paths.get(DATA_DIR, "child_before_parent.gff3");
+    private final Path url_encoding = Paths.get(DATA_DIR, "url_encoding.gff3");
+
+    @DataProvider(name = "roundTripDataProvider")
+    public Object[][] roundTripDataProvider() {
+        return new Object[][] {
+                {ensembl_human_small , 5}, {gencode_mouse_small, 7}, {ncbi_woodpecker_small, 7}, {feature_extends_past_circular_region, 1}, {with_fasta, 2}, {with_fasta_artemis, 2},
+                {ordered_cofeature, 3}, {child_before_parent, 1}, {url_encoding, 1}
+        };
+    }
+
+    @Test(dataProvider = "roundTripDataProvider")
+    public void testRoundTrip(final Path path, final int expectedCompression) {
+
+        final HashSet<String> comments1 = new HashSet<>();
+        final HashSet<SequenceRegion> regions1 = new HashSet<>();
+        final LinkedHashSet<Gff3Feature> features1 = readFromFile(path, comments1, regions1);
+
+            //write out to temp files (one gzipped, on not)
+        try {
+            final Path tempFile = Files.createTempFile("gff3Writer", ".gff3");
```

Comment:
Use one of the IOUtil.newTempPath methods to be sure the files are deleted.

Reply:
I know you want to use a dataprovider and not duplicate everything, but maybe you can factor out out a function that runs on each file instead of just doubling the code, and then call it twice here?

Reply:
done